API for object pagination, count(*)

silvanheller opened the pull request:
This PR builds upon #285, which is why it targets it so the diff is meaningful. It is focused on the use case of allowing a frontend to paginate through all elements in a large collection.

It adds the following features:
- New API Endpoint: `count/table/{table}` which is SELECT count(*) FROM table. This is needed to determine to nicely render pagination options (e.g. number of pages)
- New API Endpoint: `find/object/all/{skip}/{limit}` which basically implements SELECT * FROM multimediaobjects SKIP skip LIMIT limit.  This is used for the actual pagination.

It also does the following:
- Further cleanup of constant usage in the API by moving them to `APIConstants`
- Usage of `MediaObjectReader` with the try-with-resources syntax

sauterl commented:
Could you please also add the updated openapi specs? There seem to be missing in the changed files.

Also, in the description you write:

> New API Endpoint: count/table/table which is SELECT count(*) FROM table. This is needed to determine to nicely render pagination options (e.g. number of pages)
> New API Endpoint: find/object/all/skip/limit which basically implements SELECT * FROM multimediaobjects SKIP skip LIMIT limit. This is used for the actual pagination.

Which `table` is the parameter?

silvanheller commented:
> Could you please also add the updated openapi specs? There seem to be missing in the changed files.
> 
> Also, in the description you write:
> 
> > New API Endpoint: count/table/table which is SELECT count(*) FROM table. This is needed to determine to nicely render pagination options (e.g. number of pages)
> > New API Endpoint: find/object/all/skip/limit which basically implements SELECT * FROM multimediaobjects SKIP skip LIMIT limit. This is used for the actual pagination.
> 
> Which `table` is the parameter?

I updated the description, generated the specs, and added a "good first issue" issue to automatically generate them for a PR: https://github.com/vitrivr/cineast/issues/289

lucaro commented on cineast-core/src/main/java/org/vitrivr/cineast/core/db/DBSelector.java:
```diff
@@ -298,16 +289,32 @@
+  /**
+   * SELECT count(*) FROM table
+   */
+  default Integer rowCount(){
```
Why does this return the boxed type rather than the primitive?

sauterl commented on cineast-api/src/main/java/org/vitrivr/cineast/api/rest/handlers/actions/bool/CountRowsGetHandler.java:
```diff
@@ -0,0 +1,51 @@
+      selector.open(tableName);
+      var count = selector.rowCount();
+      LOGGER.trace("counted {} objects in table {}", count, tableName);
+      return count;
```
This is true, top-level objects have to be returned

silvanheller replied:
The method returns `Integer` and not `int`, do we still need to box it?

lucaro replied:
Yes, since Integer isn't serialized as an object but as a primitive.

lucaro commented on cineast-api/src/main/java/org/vitrivr/cineast/api/rest/handlers/actions/mediaobject/FindObjectPaginationGetHandler.java:
```diff
@@ -0,0 +1,62 @@
+    final var skipParam = parameters.get(SKIP_NAME);
+    final var skip = skipParam == null ? 0 : Integer.parseInt(skipParam);
+    final var limitParam = parameters.get(LIMIT_NAME);
+    final var limit = limitParam == null ? Integer.MAX_VALUE : Integer.parseInt(limitParam);
```
If the supplied parameter cannot be parsed to an integer, a runtime exception will be thrown which will cause an unhelpful stacktrace to be returned to the caller. The parsing should be moved into the `try/catch` block, such that you can manually return a readable error message.

silvanheller replied:
The code now returns an empty object, which is consistent with the rest of the API.
